src/ray/util/pipe_logger.cc

```diff
@@ -17,18 +17,27 @@
 #include <condition_variable>
 #include <cstring>
 #include <deque>
+#include <future>
+#include <iostream>
 #include <mutex>
 #include <string_view>
 #include <thread>
 
 #include "absl/strings/str_split.h"
-#include "ray/util/util.h"
 #include "spdlog/sinks/basic_file_sink.h"
 #include "spdlog/sinks/rotating_file_sink.h"
+#include "spdlog/sinks/stdout_color_sinks.h"
 
 namespace ray {
 
 namespace {
+
+// TODO(hjiang): Investigate how we could notify reader thread to stop by close write pipe
+// directly, instead of relying on eof indicator.
+//
+// An indicator which represents EOF, so read thread could exit.
+const std::string kEofIndicator = GenerateUUIDV4();
```

Comment:
@jjyao I still get problem when checking read return value 0, since it gets blocked forever sometimes.